### PR TITLE
Debt Ceiling Utilization

### PR DESCRIPTION
**Pull request by Vxatz**

### CODE REVIEW

File: forta-bots/Debt-Ceiling-Utilization/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# Build stage: compile Typescript to Javascript
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/Dockerfile:1]: `yarn` translation needed here

File: forta-bots/Debt-Ceiling-Utilization/src/network.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+export default class NetworkManager implements NetworkData {
+  public TeleportJoin: string;
+  public domain: string;
+  networkMap: Record<number, NetworkData>;
+
+  constructor(networkMap: Record<number, NetworkData>) {
+    this.TeleportJoin = "";
+    this.domain = "";
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/network.ts]: Use some default values which represents the real values expected

File: forta-bots/Debt-Ceiling-Utilization/src/network.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+interface NetworkData {
+  TeleportJoin: string;
+  domain: string;
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/network.ts]: should domain be a `string[]`? wdyt? In each network you will probably want to monitor multiple domains.
The requirements say:
> Domains can be either hardcoded for now

Note the plural.

File: forta-bots/Debt-Ceiling-Utilization/src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  return Finding.fromObject({
+    name: "Debt Ceiling utilization threshold exceeded",
+    description: `Debt Ceiling utilization is over ${threshold.toString()}% in TeleportJoin contract`,
+    alertId: "MK-07",
+    protocol: "MakerDAO",
+    severity: FindingSeverity.Info,
+    type: FindingType.Info,
+    metadata: {
+      debt: debt.toString(),
+      line: line.toString(),
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/utils.ts]:
```suggestion
    description: `Debt Ceiling utilization threshold exceeded in TeleportJoin contract`,
    alertId: "MK-07",
    protocol: "MakerDAO",
    severity: FindingSeverity.Info,
    type: FindingType.Info,
    metadata: {
      debt: debt.toString(),
      line: line.toString(),
      threshold: threshold.toString(),
```
Try to keep descriptions neat and add the relevant info in the metadata.

File: forta-bots/Debt-Ceiling-Utilization/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    mockGetLine.mockClear();
+    resetAllWhenMocks();
+  });
+
+  it("should return no findings when debt/line is below threshold", async () => {
+    //line, debt, blockNumber
+    const TEST_CASES: [BigNumber, BigNumber, number][] = [
+      [BigNumber.from("9387592759532123"), BigNumber.from(234), 1234],
+      [BigNumber.from("19387592759532123"), BigNumber.from(55), 5678],
+      [BigNumber.from("92759532123"), BigNumber.from(-94595), 9999999],
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/agent.spec.ts]: Is that negative `debt` intentional, does it makes sense that a debt can be negative?

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: forta-bots/Debt-Ceiling-Utilization/src/agent.spec.ts (hunk 5 above)
Comment (Vxatz) [reply forta-bots/Debt-Ceiling-Utilization/src/agent.spec.ts]: @e1Ru1o ,

Yes, according to the monitoring notion page "can be < 0 when settlement occurs before mint"

File: forta-bots/Debt-Ceiling-Utilization/src/agent.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  return async (blockEvent: BlockEvent) => {
+    const findings: Finding[] = [];
+
+    const line: BigNumber = await fetcher.getLine(data.TeleportJoin, data.domains[0], blockEvent.blockNumber);
+    const debt: BigNumber = await fetcher.getDebt(data.TeleportJoin, data.domains[0], blockEvent.blockNumber);
+
+    // Debt/Line > threshold%
+    if (debt.mul(BigNumber.from(100)).gt(line.mul(threshold))) {
+      findings.push(createFinding(debt, line, threshold));
+    }
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/agent.ts]: the idea of having a list as domain is to loop through the domains, not to just change the type and make you us `domains[0]` everywhere. Please change that, and add a test for it too

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: forta-bots/Debt-Ceiling-Utilization/package.json [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    "push": "forta-agent push",
+    "disable": "forta-agent disable",
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest --detectOpenHandles",
+    "format": "prettier --write \"src/**/*.ts\""
+  },
+  "dependencies": {
+    "forta-agent": "^0.1.6"
+  },
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/package.json]: `lru-cache` dependency missing

File: forta-bots/Debt-Ceiling-Utilization/src/network.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+const KOVAN_TESTNET_DATA: NetworkData = {
+  TeleportJoin: "0x556D9076A42Bba1892E3F4cA331daE587185Cef9",
+  domains: ["0x4b4f56414e2d534c4156452d4f5054494d49534d2d3100000000000000000000"],
+};
+
+//Arbitrum
+const RINKEBY_TESTNET_DATA: NetworkData = {
+  TeleportJoin: "0x894DB23D804c626f1aAA89a2Bc3280052e6c4750",
+  domains: ["0x52494e4b4542592d534c4156452d415242495452554d2d310000000000000000"],
+};
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/src/network.ts]: Please clarify that this networks are eth testnets, not from the comment that each of them have above

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: forta-bots/Debt-Ceiling-Utilization/README.md [hunk 9]
```diff
@@ -21,3 +21,7 @@ This bot detects when the `Debt Ceiling` utilization threshold is exceeded in `T
     - `debt`: Current TeleportJoin's debt.
     - `line`: Current TeleportJoin's line.
     - `threshold`: Configured threshold.
+
+## Test Data
+
+The bot behaviour can be verified on Polygon Mumbai Testnet (PoC).
```
Comment (e1Ru1o) [inline forta-bots/Debt-Ceiling-Utilization/README.md]: can you specify a txn to make it simple?

**Review by e1Ru1o — APPROVED** [review]

**Pull request merged**